### ISSUE TITLE
Custom Elements using `ElementInternals` to set `role` are flagged when `aria-label` is provided

### ISSUE THREAD

**Issue by DRiFTy17** [dequelabs/axe-core#4259]
### Product

axe-core

### Product Version

4.8.2

### Latest Version

- [X] I have tested the issue with the latest version of the product

### Issue Description

We are currently migrating our Web Component library to use the `ElementInternals` API to set default ARIA attributes, such as `role`, on all of our semantic host elements instead of sprouting these `role` attributes when the element is connected. This allows for setting default semantics, while still allowing for consumers to override these semantics if necessary. Upon doing so, axe is flagging _some_ of these custom elements as critical issues that appear to be false positives.

It's worth noting that axe will not flag these elements if any non-empty text nodes appear in the shadow DOM of the element, or if the `role` attribute is set directly on the element.

#### Expectation
Custom Elements that use `ElementInternals` to set their default `role` should not be flagged by axe-core.

#### Actual
The extension produced the following critical error:
> aria-label attribute cannot be used on a custom-progress-internals with no valid role attribute. 

#### How to Reproduce
The following codepen is a reproduction in the simplest form that shows a custom element with a default `role` of `progressbar` set via `ElementInternals`, along with an `aria-label` set in the light DOM on the host:
https://codepen.io/drifty17/full/WNPgvoM

Open the axe extension in DevTools and scan the page to see the error:
![image](https://github.com/dequelabs/axe-core/assets/2653457/0661e4d6-fda5-445a-b7b5-5ccf5c44f42d)

Upon viewing the accessibility tree you can also see that the semantics appear to be set properly (tested with VoiceOver as well and it interpreted it correctly):
![image](https://github.com/dequelabs/axe-core/assets/2653457/605064e6-7bd5-4eea-ab58-75b94d4deb01)

#### Additional context
I added an additional (working) example of a separate custom element for comparison that is identical except that it will sprout the `role` attribute when it connects to the DOM. When the `role` attribute is present on the element, axe does not flag the element, but when the same `role` is set via `ElementInternals` instead it does get flagged.

**Comment by straker** [comment]
Thanks for the issue. We'll need to do some extensive research to see how much assistive technologies support different element internal properties. If they support it, adding support to axe-core will take a lot of time as we have exclusively been looking at just the available element attributes to get ARIA information. Supporting this will require an extensive change in how we accesses ARIA information throughout the entire codebase.

**Comment by Westbrook** [comment]
Would love to see this added the library!!

**Comment by Westbrook** [comment]
Ooof, the move to `axe-core` being part of Lighthouse 12 makes not supporting this even more problematic. A red test (because the tool doesn't support the real world) doesn't get looked at when it gets more red, which opens to door for content to get less accessible under the guise of "but, I was using axe-core".

How could the community support bringing this feature to the project?

**Comment by straker** [comment]
@Westbrook The biggest support we could get from the community is to determine how well `ElementInternals` is supported in various assistive technologies and browsers. We've seen before that [some ATs and browsers](https://github.com/dequelabs/axe-core/issues/4085) haven't implemented the feature very well which would cause problems for users. How well it's supported would have to be a very detailed listing of different combinations of roles, uses, parent / child relationships, and common types of components (like comboboxes, forms, etc.). Once we've determined the amount of support we can determine what the next step is in terms of what axe-core should allow / report back to the user on.

**Comment by Westbrook** [comment]
Thanks for the feedback @straker! 🙇🏼 Digging into that issue a little and then will see what sort of test suite I might be able to work together to get some of this information. Is there a test context that the team uses for ensuring other accessibility tree relationships that would be good to work from, in hopes of reducing friction further down the road?

**Comment by Westbrook** [comment]
Related, @straker have you ever worked with https://playwright.dev/docs/api/class-accessibility#accessibility-snapshot? While it's deprecated (pouring on our for it), I've had pretty nice success with it. Would a test suite that tests for an equivelent accessibility tree between a native DOM implementation and a custom element implementation of similar being a useful first pass?

Playwright runs across WebKit, Firefox and Chromium and while it removes the use of multiple screen readers in the test phase the assumption that the trees being of similar shape should give enough signal, even before moving to a fully manual review of the various screen readers, right?

**Comment by Westbrook** [comment]
The above would allow us to write demos like [this](https://github.com/Westbrook/axe-core-element-internals/blob/main/stories/role.img.stories.ts) and then test their equality like [this](https://github.com/Westbrook/axe-core-element-internals/blob/main/test/role.img.test.ts). `dom` and `customElement` should create the same tree in each browser with `domNegative` confirming the test isn't accidentally giving unexpected, but matching, results.

The logs therein come out like:
```
 🚧 Browser logs on Chromium:
      {
        "role": "WebArea",
        "name": "",
        "children": [
          {
            "role": "image",
            "name": "An image from picsum.photos"
          }
        ]
      }
      {
        "role": "WebArea",
        "name": "",
        "children": [
          {
            "role": "image",
            "name": "An image from picsum.photos: DOM"
          }
        ]
      }
      {
        "role": "WebArea",
        "name": "",
        "children": [
          {
            "role": "image",
            "name": "An image from picsum.photos"
          }
        ]
      }
 🚧 Browser logs on Firefox:
      {
        "role": "document",
        "name": "",
        "children": [
          {
            "role": "img",
            "name": "An image from picsum.photos"
          }
        ]
      }
      {
        "role": "document",
        "name": "",
        "children": [
          {
            "role": "img",
            "name": "An image from picsum.photos: DOM"
          }
        ]
      }
      {
        "role": "document",
        "name": "",
        "children": [
          {
            "role": "img",
            "name": "An image from picsum.photos"
          }
        ]
      }
 🚧 Browser logs on Webkit:
      {
        "role": "WebArea",
        "name": "",
        "children": [
          {
            "role": "image",
            "name": "An image from picsum.photos"
          }
        ]
      }
      {
        "role": "WebArea",
        "name": "",
        "children": [
          {
            "role": "image",
            "name": "An image from picsum.photos: DOM"
          }
        ]
      }
      {
        "role": "WebArea",
        "name": "",
        "children": [
          {
            "role": "image",
            "name": "An image from picsum.photos"
          }
        ]
      }
```

**Comment by straker** [comment]
Thanks for that information. Unfortunately what the browser reports / logs at the accessibility tree layer is only a partial understanding. In the end what matters is what screen readers do. So it will require using screen readers to understand what they do with the information they're given.

To answer the prior question, I talked with the team and we don't have a single page where we have test cases. The closet we would have is [all our individual integration tests](https://github.com/dequelabs/axe-core/tree/develop/test/integration/rules) which shows how we expect different combinations of things to work. However they feel that before we ask for community help in testing that we'll need to come up with the testing matrix first to make sure we understand what needs to be covered before testing.

**Comment by Westbrook** [comment]
So the reason I think this approach works is that if this DOM:
```
	<div>
		<ul>
			<li>One</li>
			<li>Two</li>
			<li>Three</li>
		</ul>
		<ol>
			<li>One</li>
			<li>Two</li>
			<li>Three</li>
		</ol>
		<div role="list">
			<div role="listitem">One</div>
			<div role="listitem">Two</div>
			<div role="listitem">Three</div>
		</div>
	</div>
```
And this DOM:
```
	<style>
		/**
		* When applied via ElementInternals, Firefox adds the "list item marker" to the tree.
		* This applies them in a way that is equivelent to the native elements.
		* An in project implemenation would likely want to apply this via shadow DOM, adoptedStyleSheets, and a <slot> element.
		*/ 
		axe-list {
			list-style-type: none;
		}
		axe-list[unordered] {
			list-style-type: disc;
		}
		axe-list[ordered] {
			list-style-type: decimal;
		}
		axe-list-item {
			display: list-item;
		}
	</style>
	<div>
		<axe-list unordered>
			<axe-list-item>One</axe-list-item>
			<axe-list-item>Two</axe-list-item>
			<axe-list-item>Three</axe-list-item>
		</axe-list>
		<axe-list ordered>
			<axe-list-item>One</axe-list-item>
			<axe-list-item>Two</axe-list-item>
			<axe-list-item>Three</axe-list-item>
		</axe-list>
		<axe-list>
			<axe-list-item>One</axe-list-item>
			<axe-list-item>Two</axe-list-item>
			<axe-list-item>Three</axe-list-item>
		</axe-list>
	</div>
```
were to both create the same tree, it no longer matters what the screen reader does with with the accessibility tree. Either the screen readers already leverage DOM incorrectly or they will do the same things as the native elements. I understand that it isn't the same as manual testing, but I do see it as having a high quality signal as to the direction this work should go in. However, if that's still not a valid enough step, then we can work towards a different solution.

Having a testing matrix in either case will be very helpful. Would you and the team be able to prioritize preparing something like that? I'm the chairperson at https://github.com/w3c/webcomponents-cg and I'd love to get a concrete path forward that we could spread out across a number of accessibility interested members of the group (which should be all of them, 😉).

**Comment by WilcoFiers** [comment]
As far as I can tell there is no reliable way to get to the `ElementInternals` object. First up, ElementInternals may not be saved as a prop on the element, or it may be a private prop. Axe-core won't ever be able to get to it if that's the case.

Even if `ElementInternals` is a string prop on the element, axe won't know what that is. The only way to have a chance of finding it is to loop over all props of the element, and check if any of them are an `ElementInternals` object. We can't just read those properties out either. There may be getters o that element. Axe can't just invoke a whole bunch of getters. That can result in all sorts of destructive things. `Object.getOwnPropertyDescriptor()` can help us figure out what's a getter and what isn't.

There are a lot of opportunities for things to go wrong here. And if axe can't detect an attached `ElementInternals` object, you are opening the floodgate to false positives. I'm more inclined to start a conversation with the standards folks to try and get an API for accessing `ElementInternals`. Without that I think this is far too much of a foot gun. I'd be more inclined to explicitly not support `ElementInternals` in axe-core and recommend against using it (or at least, its accessibility props). As much as I'd like for axe-core to support this, if it can't be done well, I'm hesitant to try and do it at all.

**Comment by Westbrook** [comment]
While the complexity of possibly expanding browser APIs moves forward...

Thinking about this more, I wonder if there is a configuration point that I've not yet found wherein a consuming developer could expand the values exported by https://github.com/dequelabs/axe-core/blob/develop/lib/commons/standards/implicit-html-roles.js#L72? In this way, while a developer would be taking on some level of responsibility for the truth of the role applied to an element, the wide ranging benefits of `axe-core` could still largely be leveraged across a project.

**Comment by bennypowers** [comment]
@WilcoFiers can you patch attachInternals?

https://codepen.io/bennyp/pen/RwzaWwd

```js
const aXeMap = new WeakMap;
const orig = HTMLElement.prototype.attachInternals; HTMLElement.prototype.attachInternals = function () {
  const internals = orig.call(this);
  aXeMap.set(this, internals);
  return internals;
};
```

**Comment by bennypowers** [comment]
There are cases where aXe-core informs users that automated tests are insufficient and manual testing is required.

Being as how this feature is widely available and used in production, shouldn't aXe-core surface cases like these are warnings instead of errors?

**Comment by WilcoFiers** [comment]
I've opened an issue with WHATWG here: [.elementInternals accessor property whatwg/html#11040](https://github.com/whatwg/html/issues/11040). People who upvoted here, please go there and leave your feedback as well. 

Unfortunately I don't think there is a way around this. **Without new DOM APIs, the only way to avoid false positives on pages with ElementInternals is for axe-core to stop testing custom elements all together.** Even the testing of native elements will be affected by this since this affects the computation of accessible names. Scenarios like the following can not be failed:

```html
<label>
  <my-elm></my-elm>
  <input type="text" name="is-my-label-empty" />
</label>
```

There are work-around through some browser extensions and web drivers, but axe-core cannot work that way without giving up on the third item of its [manifesto](https://github.com/dequelabs/axe-core?tab=readme-ov-file#about-axe---our-manifesto): work on all modern browsers and with whatever tools, frameworks, libraries and environments.

I don't see that happening. More likely we'll create new tools for this in the axe Extension. That'll work, but that isn't going to solve anything for Lighthouse, Accessibility Insights, etc.

**Comment by Westbrook** [comment]
In lieu of https://github.com/whatwg/html/issues/11040 would a [community protocol](https://github.com/webcomponents-cg/community-protocols) of the same capability be something that the main library would be amenable to supporting? It would be a sort of chicken or the egg bit of development, without agreement that it could come under use we'd likely not get enough involvement for it to be use actual use, but agreement in user code that an `axe-core` testable element internals looks like X could be devised, even as the platform itself endeavored to close over access of this data. Were you and the team open to something like this, I know that the [Web Components Community Group](https://github.com/w3c/webcomponents-cg) would be open to taking something like this on with earnest. In fact, @davatron5000 brought up something in this area during our December meeting, so the conversation wouldn't need to start from zero.

**Comment by bennypowers** [comment]
Red Hat would be interested in implementing such a community protocol in lieu of / while waiting for browser support

**Comment by WilcoFiers** [comment]
@Westbrook @bennypowers Thank you for the suggestions. We're working in the background to try and figure out a viable path forward for this. We'll try to come up with an answer, but it may take us time. A community protocol could be a useful way for axe to learn about Internals if they are used, but the bigger problem we're trying to find a solution for is to detect when an element doesn't have internals so we don't need to skip it for accessibility testing.

**Comment by straker** [comment]
Some more info, [jsdom v26](https://github.com/jsdom/jsdom/blob/main/Changelog.md#2600) added preliminary support for `ElementInternals` `label` property.

**Comment by WilcoFiers** [comment]
I was looking at `formAssociated` today (#4750), where I ran into trouble trying to access it from a different javascript world. I.e if a host script sets `formAssociated` to `true`, chrome extensions, playwright, puppeteer, etc., because they run in a different world do not have access to that `formAssociated` property. 

That limitation has implications for this discussion too that I hadn't realized before. If axe runs in a different JS world it won't have access to any properties on the custom element. A community protocol where everyone sets its ElementInternals on a `.internals` property won't help because axe won't have access to it.

**Comment by aarongustafson** [comment]
> A community protocol where everyone sets its `ElementInternals` on a `.internals` property won't help because axe won't have access to it.

As I understand it, Axe would have access to it as long as it knew the appropriate anchor point (property name). So in the example at the start of this thread, knowing the prop is `internals` would allow you to inspect the accessibility plumbing (you can test this using the console on the linked CodePen demo)


```js
const $foo = document.querySelector("custom-progress-internals");
$foo.internals; // ElementInternals object
```

Most of the docs (MDN, etc.) favor `_internals`, but community agreement on this could be useful and Axe could look for that prop by default.

It may also make sense to use a configuration file in the `.well-known` directory (or similar) to enable site-wide customization. That would at least cover sites where they are working with a specific WC-based design system.

**Comment by bennypowers** [comment]
Maybe a global weakmap would be palatable? 

```ts
globalThis._elementInternals = new WeakMap<Element, ElementInternals>;
```

This has the advantage of not changing the class shape. It's common practise to wrap attachInternals with some abstraction, in which case this could be implemented in one place easily enough.

```ts
// class shape doesn't change
// so you can keep your #
#internals = new InternalsController(this); 
```

```ts
class InternalsController {
  constructor(host: Element) {
    this.#internals = host.attachInternals();
    globalThis._elementInternals.set(host, this.#internals);
  } 
}
```

**Comment by dbjorge** [comment]
> As I understand it, Axe would have access to it as long as it knew the appropriate anchor point (property name).

Unfortunately, I think you are misunderstanding Wilco's point. Axe needs to support running in what browsers call an ["isolated world"](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts#isolated_world), a separate JavaScript execution context from the "main" execution context of the page but which shares access to the page's DOM. The most common example (and one that Axe specifically needs to support) is webextension Content Scripts. A script running in one isolated world cannot see arbitrary properties added to DOM nodes (or to `globalThis`) by a different isolated world.

**Comment by aarongustafson** [comment]
> Unfortunately, I think you are misunderstanding Wilco's point. Axe needs to support running in what browsers call an ["isolated world"](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts#isolated_world), a separate JavaScript execution context from the "main" execution context of the page but which shares access to the page's DOM. The most common example (and one that Axe specifically needs to support) is webextension Content Scripts. A script running in one isolated world cannot see arbitrary properties added to DOM nodes (or to `globalThis`) by a different isolated world.

Gotcha. In the Playwright context, it looks like you can access the internals directly using `page.evaluate()` but from my experiments you can’t seem to extract the `ElementInternals` object directly and, instead, it seems to require mapping it over to a generic object for export. I also noted that extraction process can throw an `Error` (for example on accessing `form` when it’s not set) which is why I’ve got the `try…catch` in there.

This is working for me locally (with a bit of debugging). Note that the `authentication_hash` query string in the URL will likely need updating if you run this code yourself:

```js
test('Codepen elementInternals extraction', async ({ page }) => {
    await page.goto('https://cdpn.io/pen/debug/GggVryB?authentication_hash=DqrDdQvDeqqr');

    const SELECTOR = 'custom-progress-internals'
    const $el = page.locator('custom-progress-internals');
    await $el.waitFor({state:"attached"});

    const extracted_internals = await page.evaluate((selector) => {
      let internals = null;
      let internals_obj = null;
      
      const $el = document.querySelector(selector);
      if ( ! $el ) return;

      // I’ve seen a few different conventions for storing internals
      switch (true) {
        case "internals" in $el:
          internals = $el.internals;
          break;
        case "_internals" in $el:
          internals = $el._internals;
          break;
        case "internals_" in $el:
          internals = $el.internals_;
          break;
      }
      
      if ( internals ) {
        internals_obj = {};
        for ( let key in internals ){
          try {
            internals_obj[key] = internals[key] || undefined;
          } catch (e) {}
        }
      }

      return internals_obj;
    }, SELECTOR );

    console.log( extracted_internals );
 });
```

I know the browser extension is isolated, but I _think_ the same may not be true in the DevTools context as the console — at least — has access to the `ElementInternals` information. And I think developers loading axe-core into the page manually (via `script` would also allow it access to the necessary info). So it may just be the browser extension that is restricted in this way.

**Comment by bennypowers** [comment]
The [Web Components Community Group Protocol](https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/element-internals-declaration.md) for exposing internals to tools like aXe core has been updated and upgraded to Draft status

**Comment by padmavemulapati** [comment]
devTask

**Comment by padmavemulapati** [comment]
```
<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>axe-core issue 4259 repro</title>
</head>
<body>
  <custom-progress-internals aria-label="Loading progress"></custom-progress-internals>
  <custom-progress-internals aria-label="Internals progressbar"></custom-progress-internals>
<br/>
<custom-progress-host aria-label="Host attribute progressbar"></custom-progress-host>


  <script>
    class CustomProgressInternals extends HTMLElement {
      constructor() {
        super();
        this._internals = this.attachInternals();
        this._internals.role = 'progressbar';
      }
    }

    customElements.define('custom-progress-internals', CustomProgressInternals);
  </script>
</body>
</html>
```

QA Steps: 
1. Opened a local repro page for the above snippet,  with a custom element using ElementInternals to set role="progressbar" and aria-label on the host
2.  enable the feature flag before axe run - `axe._enableElementInternals = true;`
3.  then 
 
```
 axe.run(document, {  runOnly: {    type: 'tag',
    values: ['next']
  }
}).then(results => {
  console.log(results);
});

```

Executed await axe.run(); in DevTools Console
Expected

No violation stating the custom element has no valid role attribute on aria-label
No false positive for aria-label on the custom element


<img width="1414" height="798" alt="Image" src="https://github.com/user-attachments/assets/ec35fee8-074a-4515-b04a-0e012d6ba8d1" />

## Environment:
| **Label**                         | **Value** |
| -------------------------- | ------ |
| **Product**                | _axe-core_ |
| **version**           | _@next_ |
| **OS**                              | _Windows 11 & MAC-26.3_ |
| **Tested Browsers**       | _Chrome Version 148_ |
